### PR TITLE
NEW: QMainWindow example

### PR DESCRIPTION
**Pull request by JulienMaille**
With some proper handling of maximized mode (icon and frame size)

### CODE REVIEW

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:188]: Maybe this should be QMainWindow::paintEvent(event);

File: examples/qmainwindow/main.cpp
Comment (wangwenx190) [inline examples/qmainwindow/main.cpp:113]: Maybe this can be done in the QtAcrylicMainWindow's changeEvent?

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:25]: Use lower-case files to get the maximum portability.

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:147]: Can't these attributes be reset to default values if the user choose to turn off the acrylic blur?

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:194]: I think we don't need this signal. If you want to re-layout, just do it here.

File: qtacrylicmainwindow.h
Comment (wangwenx190) [inline qtacrylicmainwindow.h:28]: #include <QtWidgets/qmainwindow.h> -> lower-case file name

File: qtacrylicwidget.cpp
Comment (wangwenx190) [inline qtacrylicwidget.cpp:175]: Why was this needed for QtWidget? It dones't have the layout issue as QMainWindow.

File: .gitignore
Comment (wangwenx190) [inline .gitignore:72]: What's this for?

**Review by wangwenx190 — CHANGES_REQUESTED** [review]
1. You can add your own copyright notice in the file headers if you want.
2. Looks like some modifications are needed before merging.
3. I'm not sure whether this PR is finished or not. If you think it's ready for merging, just tell me.

File: qtacrylicmainwindow.cpp
Comment (JulienMaille) [reply qtacrylicmainwindow.cpp:188]: sure

File: .gitignore
Comment (JulienMaille) [reply .gitignore:72]: don't you have many .res files generated by msvc?

Comment (wangwenx190) [reply .gitignore:72]: MSVC will generate many .res files indeed, but they won't pullote your source code dir if you are doing shadow builds (out-of-source builds). Qt Creator will always do shadow builds by default. These files will be in a separate dir. Maybe you are doing in-source builds, and thus you have these files in the source directory.

File: qtacrylicmainwindow.cpp
Comment (JulienMaille) [reply qtacrylicmainwindow.cpp:147]: why would the default values be the right ones? I don't think toggling acrylic post startup is someone a user would do

File: qtacrylicmainwindow.cpp
Comment (JulienMaille) [reply qtacrylicmainwindow.cpp:194]: I can mode the setContentMargin() here but the toolbar related stuff has to stay in the example implementation I think

File: .gitignore
Comment (JulienMaille) [reply .gitignore:72]: could be, is that a problem to have this in the .ignore file?

**Comment by JulienMaille** [comment]
@wangwenx190 have a look at latest commits

File: .gitignore
Comment (wangwenx190) [reply .gitignore:72]: No, just point this out. I'm fine with it.

**Comment by wangwenx190** [comment]
Looks nice now. Is it finished? If there are no more commits, I'll merge it.

**Comment by JulienMaille** [comment]
Have you tried to compile it on your side?

**Comment by wangwenx190** [comment]
> 
> 
> Have you tried to compile it on your side?

No I haven't. I just read the source code. But it's OK. It looks good in general. I'll merge it if you think it's ready. If some issues are found after merging, we can fix them afterwards.

**Review by JulienMaille — COMMENTED** [review]
I'm good

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:147]: No need to disconnect, because the window frame is also draw by the acrylic helper. This signal is used by both the acrylic background and the window frame.

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:171]: `showEvent()` will be executed everytime the window shows, so these functions will be called many times, you have to add a guard to make sure they are executed only once, such as move them to the `if (!m_inited) {` block.

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [inline qtacrylicmainwindow.cpp:176]: you can use `devicePixelRatioF()` to replace `windowHandle()->devicePixelRatio()`, because windowHandle() maybe null

**Review by wangwenx190 — COMMENTED** [review]
Looks good in general, just some minor issues.

File: qtacrylicmainwindow.cpp
Comment (wangwenx190) [reply qtacrylicmainwindow.cpp:147]: And due to we will never disconnect, you should not make the connection here. Maybe move it to the showEvent is more appropriate.

Comment (JulienMaille) [reply qtacrylicmainwindow.cpp:147]: the frame is drawn in the `paintEvent()` why would you need extra repaints?

Comment (wangwenx190) [reply qtacrylicmainwindow.cpp:147]: the repaint signal is emited when the window state and activation changes, we need to re-draw the window frame in that case because the frame color need to be changed.

Comment (wangwenx190) [reply qtacrylicmainwindow.cpp:147]: the paintEvent won't be triggered if the window state or activation changes (because Qt think the window doesn't need to repaint), we need to trigger it manually.

File: examples/qmainwindow/main.cpp
Comment (wangwenx190) [inline examples/qmainwindow/main.cpp:108]: Why not just use `mainWindow->winId()` or `mainWindow->windowHandle()->winId()`?

**Review by wangwenx190 — COMMENTED** [review]
Thanks!

**Comment by wangwenx190** [comment]
I have an idea, I can move the TrackPopupMenu code to the Utilities code of FramelessHelper, and you can just call it with a proper HWND and coordinate.

**Comment by wangwenx190** [comment]
I've added a `displaySystemMenu()` function in the `Utilities` namespace, you can use that to simplify your code.

File: qtacrylicmainwindow.cpp
Comment (JulienMaille) [reply qtacrylicmainwindow.cpp]: done

**Comment by JulienMaille** [comment]
It's ready for merge I believe

**Comment by wangwenx190** [comment]
Thanks!

**Pull request merged**
